File: go/store/memory/memory.go
// Package memory is the in-process store (REQ-CORE-6) for tests and single-instance deployments.
package memory

import (
	"context"
	"sync"
	"time"

	"github.com/sns45/anyonce/go/anyonce"
)

// Store keeps records in a map guarded by one mutex, so Begin is one atomic claim (D4).
type Store struct {
	mu      sync.Mutex
	records map[string]*anyonce.Record
}

// New returns an empty store.
func New() *Store { return &Store{records: make(map[string]*anyonce.Record)} }

func mapKey(scope, key string) string { return scope + "\x00" + key }

// Len returns the number of stored records, expired or not.
func (s *Store) Len() int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return len(s.records)
}

// Begin applies the 3.2 precedence: TTL expiry (absent), then fingerprint, then lease; fence continues from a
// stale row (Q8).
func (s *Store) Begin(ctx context.Context, op anyonce.Operation, opts anyonce.BeginOptions) (anyonce.BeginOutcome, error) {
	if err := ctx.Err(); err != nil {
		return anyonce.BeginOutcome{}, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	k := mapKey(op.Scope, op.Key)
	existing := s.records[k]
	now := opts.Now
	if existing != nil && existing.ExpiresAt.After(now) {
		if existing.Fingerprint != op.Fingerprint {
			rec := existing.Clone()
			return anyonce.BeginOutcome{Kind: anyonce.BeginMismatch, Record: &rec}, nil
		}
		if existing.State == anyonce.StateCompleted {
			rec := existing.Clone()
			return anyonce.BeginOutcome{Kind: anyonce.BeginCompleted, Record: &rec}, nil
		}
		if existing.LeaseUntil.After(now) {
			return anyonce.BeginOutcome{Kind: anyonce.BeginInFlight, LeaseUntil: existing.LeaseUntil}, nil
		}
	}
	var fence int64 = 1
	if existing != nil {
		fence = existing.Fence + 1
	}
	s.records[k] = &anyonce.Record{
		Scope: op.Scope, Key: op.Key, Fingerprint: op.Fingerprint, State: anyonce.StateInFlight, Fence: fence,
		LeaseUntil: now.Add(opts.Lease), CreatedAt: now, ExpiresAt: now.Add(opts.TTL),
	}
	return anyonce.BeginOutcome{Kind: anyonce.BeginAcquired, Fence: fence}, nil
}

// Complete stores the result (or the omitted form) when the fence matches.
func (s *Store) Complete(ctx context.Context, op anyonce.Operation, fence int64, result anyonce.StoredResult, now time.Time) (anyonce.CompleteStatus, error) {
	if err := ctx.Err(); err != nil {
		return "", err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	existing := s.records[mapKey(op.Scope, op.Key)]
	if existing == nil || !existing.ExpiresAt.After(now) {
		return anyonce.CompleteNotFound, nil
	}
	if existing.Fence != fence {
		return anyonce.CompleteStaleFence, nil
	}
	if existing.State == anyonce.StateCompleted {
		return anyonce.CompleteOK, nil
	}
	stored := result.Clone()
	if result.Omitted {
		stored.Body = nil
		existing.ResultOmitted = true
	}
	existing.State = anyonce.StateCompleted
	existing.Result = &stored
	return anyonce.CompleteOK, nil
}

// Abandon deletes an in-flight record when the fence matches.
func (s *Store) Abandon(ctx context.Context, op anyonce.Operation, fence int64) (anyonce.CompleteStatus, error) {
	if err := ctx.Err(); err != nil {
		return "", err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	k := mapKey(op.Scope, op.Key)
	existing := s.records[k]
	if existing == nil || existing.State != anyonce.StateInFlight {
		return anyonce.CompleteNotFound, nil
	}
	if existing.Fence != fence {
		return anyonce.CompleteStaleFence, nil
	}
	delete(s.records, k)
	return anyonce.CompleteOK, nil
}

// Get returns a copy of a live record, or nil.
func (s *Store) Get(ctx context.Context, scope, key string, now time.Time) (*anyonce.Record, error) { //nolint:nilnil // nil record means absent
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	existing := s.records[mapKey(scope, key)]
	if existing == nil || !existing.ExpiresAt.After(now) {
		return nil, nil
	}
	rec := existing.Clone()
	return &rec, nil
}

// Purge deletes expired records and returns how many.
func (s *Store) Purge(ctx context.Context, now time.Time) (int, error) {
	if err := ctx.Err(); err != nil {
		return 0, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	removed := 0
	for k, rec := range s.records {
		if !rec.ExpiresAt.After(now) {
			delete(s.records, k)
			removed++
		}
	}
	return removed, nil
}

// PhysicallyRemove is a test helper simulating a native TTL sweep, so the next Begin restarts the fence at 1.
func (s *Store) PhysicallyRemove(ctx context.Context, scope, key string) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.records, mapKey(scope, key))
	return nil
}

File: packages/core/src/memory.ts
import type {
  BeginOptions,
  BeginOutcome,
  CompleteStatus,
  IdempotencyRecord,
  OmittedResult,
  Operation,
  Store,
  StoredResult,
} from './types';
import { isOmitted } from './types';

function mapKey(op: Pick<Operation, 'scope' | 'key'>): string {
  return `${op.scope}\x00${op.key}`;
}

function copyResult(result: StoredResult): StoredResult {
  const out: StoredResult = { kind: result.kind };
  if (result.status !== undefined) out.status = result.status;
  if (result.headers !== undefined) out.headers = result.headers.map(([n, v]) => [n, v]);
  if (result.body !== undefined) out.body = new Uint8Array(result.body);
  if (result.outcome !== undefined) out.outcome = result.outcome;
  if (result.error !== undefined)
    out.error = { name: result.error.name, message: result.error.message };
  return out;
}

function omittedToStored(result: OmittedResult): StoredResult {
  const out: StoredResult = { kind: result.kind };
  if (result.status !== undefined) out.status = result.status;
  if (result.headers !== undefined) out.headers = result.headers.map(([n, v]) => [n, v]);
  return out;
}

function copyRecord(record: IdempotencyRecord): IdempotencyRecord {
  const out: IdempotencyRecord = {
    scope: record.scope,
    key: record.key,
    fingerprint: record.fingerprint,
    state: record.state,
    fence: record.fence,
    leaseUntil: record.leaseUntil,
    createdAt: record.createdAt,
    expiresAt: record.expiresAt,
  };
  if (record.result !== undefined) out.result = copyResult(record.result);
  if (record.resultOmitted !== undefined) out.resultOmitted = record.resultOmitted;
  return out;
}

/**
 * REQ-CORE-6: in-process store for tests and single-instance deployments. begin performs its read and write
 * without yielding, so concurrent callers in one event loop see one atomic claim (D4).
 */
export class MemoryStore implements Store {
  private readonly records = new Map<string, IdempotencyRecord>();

  get size(): number {
    return this.records.size;
  }

  async begin(op: Operation, opts: BeginOptions): Promise<BeginOutcome> {
    const key = mapKey(op);
    const existing = this.records.get(key);
    const { now } = opts;
    if (existing !== undefined && existing.expiresAt > now) {
      if (existing.fingerprint !== op.fingerprint)
        return { outcome: 'mismatch', record: copyRecord(existing) };
      if (existing.state === 'completed')
        return { outcome: 'completed', record: copyRecord(existing) };
      if (existing.leaseUntil > now)
        return { outcome: 'in_flight', leaseUntil: existing.leaseUntil };
    }
    const fence = existing === undefined ? 1 : existing.fence + 1;
    this.records.set(key, {
      scope: op.scope,
      key: op.key,
      fingerprint: op.fingerprint,
      state: 'in_flight',
      fence,
      leaseUntil: now + opts.leaseMs,
      createdAt: now,
      expiresAt: now + opts.ttlMs,
    });
    return { outcome: 'acquired', fence };
  }

  async complete(
    op: Operation,
    fence: number,
    result: StoredResult | OmittedResult,
    now: number,
  ): Promise<CompleteStatus> {
    const existing = this.records.get(mapKey(op));
    if (existing === undefined || existing.expiresAt <= now) return 'not_found';
    if (existing.fence !== fence) return 'stale_fence';
    if (existing.state === 'completed') return 'ok';
    existing.state = 'completed';
    if (isOmitted(result)) {
      existing.result = omittedToStored(result);
      existing.resultOmitted = true;
    } else {
      existing.result = copyResult(result);
    }
    return 'ok';
  }

  async abandon(op: Operation, fence: number): Promise<CompleteStatus> {
    const key = mapKey(op);
    const existing = this.records.get(key);
    if (existing === undefined || existing.state !== 'in_flight') return 'not_found';
    if (existing.fence !== fence) return 'stale_fence';
    this.records.delete(key);
    return 'ok';
  }

  async get(op: Pick<Operation, 'scope' | 'key'>, now: number): Promise<IdempotencyRecord | null> {
    const existing = this.records.get(mapKey(op));
    if (existing === undefined || existing.expiresAt <= now) return null;
    return copyRecord(existing);
  }

  async purge(now: number): Promise<number> {
    let removed = 0;
    for (const [key, record] of this.records) {
      if (record.expiresAt <= now) {
        this.records.delete(key);
        removed += 1;
      }
    }
    return removed;
  }

  /** Test-only: simulates a native TTL sweep deleting the row, so the next begin restarts the fence at 1 (Q8). */
  async physicallyRemove(op: Pick<Operation, 'scope' | 'key'>): Promise<void> {
    this.records.delete(mapKey(op));
  }
}
